### ISSUE TITLE
Home Connect API token expired

### ISSUE THREAD

**Issue by bje** [DavidMStraub/homeassistant-homeconnect#17]
Over the weekend I started getting this error:

`2019-03-12 20:53:00 ERROR (MainThread) [homeassistant.setup] Error during setup of component homeconnect
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 395, in request
    http_method=method, body=data, headers=headers)
  File "/srv/homeassistant/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 198, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/setup.py", line 154, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/srv/homeassistant/config/custom_components/homeconnect/__init__.py", line 72, in setup
    hass.data[DOMAIN]['devices'] = get_devices(hc)
  File "/srv/homeassistant/config/custom_components/homeconnect/__init__.py", line 129, in get_devices
    appl = retry(hc.get_appliances, times=20, sleep=0.1)
  File "/srv/homeassistant/config/custom_components/homeconnect/__init__.py", line 120, in retry
    return f()
  File "/srv/homeassistant/lib/python3.6/site-packages/homeconnect/api.py", line 146, in get_appliances
    data = self.get(ENDPOINT_APPLIANCES)
  File "/srv/homeassistant/lib/python3.6/site-packages/homeconnect/api.py", line 101, in get
    res = self.oauth.get(uri)
  File "/srv/homeassistant/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/srv/homeassistant/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 408, in request
    self.auto_refresh_url, auth=auth, **kwargs
  File "/srv/homeassistant/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 374, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/srv/homeassistant/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 415, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/srv/homeassistant/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 425, in parse_token_response
    validate_token_parameters(params)
  File "/srv/homeassistant/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 432, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/srv/homeassistant/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 405, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) invalid refresh_token`

I've created a new app in Home Connect Developers portal and tried using new credentials, I still get the same error.

Versions in use:
`(homeassistant) homeassistant@hass0:~$ pip freeze | egrep -i '^homeconnect=|^homeassistant='
homeassistant==0.89.1
homeconnect==0.2.1`

I'm also using the master branch of`DavidMStraub/homeassistant-homeconnect` and this is located in `custom_components/homeconnect`.

Any thoughts?

**Comment by DavidMStraub** [comment]
As a workaround you can delete the `.homeconnect-token-cache` in your Home Assistant config directory. It also happened to me once and I am not sure what is causing it.

**Comment by bje** [comment]
That worked, thank you so much! 👍

**Comment by sharkdp** [comment]
After an update of home-assistant to 0.102.1, I get the same error, but I don't have a `.homeconnect-token-cache` file in my `/config` folder (everything worked fine before).

I am using v0.2.1 (installed via HACS) of homeassistant-homeconnect.

(PS: thank you very much for this integration!)

**Comment by DavidMStraub** [comment]
In the new version, the tokens (including refreshing them) is handled by Home Assistant itself, so I'm surprised, not sure the integration can do something about that. Did you add the Home Connect integration using the "integrations" page?

**Comment by sharkdp** [comment]
The specific backtrace that I get is slightly different than the one posted above:
```
Error setting up entry Configuration.yaml for homeconnect

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/homeconnect/api.py", line 65, in request
    return getattr(self._oauth, method)(url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests_oauthlib/oauth2_session.py", line 478, in request
    url, http_method=method, body=data, headers=headers
  File "/usr/local/lib/python3.7/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 198, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 192, in async_setup
    hass, self
  File "/config/custom_components/homeconnect/__init__.py", line 82, in async_setup_entry
    await update_all_devices(hass, entry)
  File "/config/custom_components/homeconnect/__init__.py", line 114, in update_all_devices
    data[DEVICES] = await hass.async_add_executor_job(hc.get_devices)
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/homeconnect/api.py", line 45, in get_devices
    appl = self.get_appliances()
  File "/usr/local/lib/python3.7/site-packages/homeconnect/api.py", line 123, in get_appliances
    data = self.get(ENDPOINT_APPLIANCES)
  File "/usr/local/lib/python3.7/site-packages/homeconnect/api.py", line 73, in get
    res = self.request("get", endpoint)
  File "/usr/local/lib/python3.7/site-packages/homeconnect/api.py", line 67, in request
    self._oauth.token = self.refresh_tokens()
  File "/config/custom_components/homeconnect/api.py", line 39, in refresh_tokens
    self.session.async_ensure_token_valid(), self.hass.loop
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 403, in async_ensure_token_valid
    new_token = await self.implementation.async_refresh_token(self.token)
AttributeError: 'OAuth2Session' object has no attribute 'async_refresh_token'
```

I installed home-connect via HACS and then added it via "Integrations", yes. I still have the
```yaml
homeconnect:
  client_id: !secret homeconnect_id
  client_secret: !secret homeconnect_secret
```
entry in my `configuration.yaml`, though.

I previously had the custom component manually installed without HACS, but I completely removed the folder in `/config/custom_components/` before reinstalling via HACS.

**Comment by sharkdp** [comment]
I removed and added the integration and now it works again. I will report back if this error reappears.

**Comment by DavidMStraub** [comment]
Thanks for reporting this. I also just run into it and opened a new issue #43.

**Comment by jhqv** [comment]
Just reporting in... 
The same thing happened to me. I successfully added the Home Connect integration last week, and it worked fine. A few days ago I discovered that the entities were no longer available, and I found the "token expired" message in the log. Removing the integration and re-adding it solved the issue for me too.

(Problem started on Home Assistant version 0.110.2, upgrade to 0.110.3 didn't help.)
